src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -44,71 +43,75 @@ internal static BlobBuilder MethodSignatureEncoder(ModuleBuilderImpl module, Typ
             }
 
             return methodSignature;
         }
 
-        private static void WriteSignatureTypeForReflectionType(SignatureTypeEncoder signature, Type type, ModuleBuilderImpl module)
+        private static void WriteSignatureForType(SignatureTypeEncoder signature, Type type, ModuleBuilderImpl module, Action typedReference)
         {
             CoreTypeId? typeId = module.GetTypeIdFromCoreTypes(type);
 
-            // We need to translate from Reflection.Type to SignatureTypeEncoder.
             switch (typeId)
             {
                 case CoreTypeId.Boolean:
                     signature.Boolean();
-                    break;
+                    return;
                 case CoreTypeId.Byte:
                     signature.Byte();
-                    break;
+                    return;
                 case CoreTypeId.SByte:
                     signature.SByte();
-                    break;
+                    return;
                 case CoreTypeId.Char:
                     signature.Char();
-                    break;
+                    return;
                 case CoreTypeId.Int16:
                     signature.Int16();
-                    break;
+                    return;
                 case CoreTypeId.UInt16:
                     signature.UInt16();
-                    break;
+                    return;
                 case CoreTypeId.Int32:
                     signature.Int32();
-                    break;
+                    return;
                 case CoreTypeId.UInt32:
                     signature.UInt32();
-                    break;
+                    return;
                 case CoreTypeId.Int64:
                     signature.Int64();
-                    break;
+                    return;
                 case CoreTypeId.UInt64:
                     signature.UInt64();
-                    break;
+                    return;
                 case CoreTypeId.Single:
                     signature.Single();
-                    break;
+                    return;
                 case CoreTypeId.Double:
                     signature.Double();
-                    break;
+                    return;
                 case CoreTypeId.IntPtr:
                     signature.IntPtr();
-                    break;
+                    return;
                 case CoreTypeId.UIntPtr:
                     signature.UIntPtr();
-                    break;
+                    return;
                 case CoreTypeId.Object:
                     signature.Object();
-                    break;
+                    return;
                 case CoreTypeId.String:
                     signature.String();
-                    break;
-                default:
-                    throw new NotSupportedException(SR.Format(SR.NotSupported_Signature, type.FullName));
+                    return;
+                case CoreTypeId.TypedReference:
+                    typedReference();
```

Comment:
This should be just `signature.WriteTypeCode(SignatureTypeCode.TypedReference)`. This callback should not be needed.

Reply:
Yeah I tried that, it was private, and I did not find any other way to write `TypedReference` from `SignatureTypeEncoder`, that is why I end up using this callback https://github.com/dotnet/runtime/blob/2087ceb0c70f0c13e79fbbdec90efd242e1f2cd4/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs#L897

Reply:
You can do `signature.Builder.WriteByte((byte)SignatureTypeCode.TypedReference))` then - it should be all public if I am reading it correctly.

Reply:
Right, thanks